File: clojure/robot-simulator/src/robot_simulator.clj
(ns robot-simulator)

(defn robot [coordinates bearing]
  {:coordinates coordinates :bearing bearing})

(defn turn-right [bearing]
  (case bearing
    :north :east
    :east :south
    :south :west
    :west :north))

(defn turn-left [bearing]
  (case bearing
    :north :west
    :east :north
    :south :east
    :west :south))

(defn- letter->action [letter]
  (case letter
    \A :advance
    \L :turn-left
    \R :turn-right))

(defn- update-position [{:keys [coordinates bearing]} delta]
  (let [new-positon (merge-with + coordinates delta)]
    (robot new-positon bearing)))

(defn- move-left [{:keys [coordinates bearing]}]
  (robot coordinates (turn-left bearing)))

(defn- move-right [{:keys [coordinates bearing]}]
  (robot coordinates (turn-right bearing)))

(defn- advance [bot]
  (let [bearing (:bearing bot)]
    (case bearing
      :north (update-position bot {:x 0  :y  1})
      :south (update-position bot {:x 0  :y -1})
      :east  (update-position bot {:x 1  :y  0})
      :west  (update-position bot {:x -1 :y  0}))))

(defn- move [bot action]
  (case action
    :advance (advance bot)
    :turn-left (move-left bot)
    :turn-right (move-right bot)))

(defn simulate [actions bot]
  (reduce move bot (map letter->action actions)))

File: ruby/robot-simulator/robot_simulator.rb
class Robot
  attr_reader :bearing, :coordinates

  BEARINGS = %i[north east south west]

  def turn_right
    current_index = BEARINGS.find_index(bearing)
    next_index = (current_index + 1) % BEARINGS.size
    next_bearing = BEARINGS[next_index]

    orient(next_bearing)
  end

  def turn_left
    current_index = BEARINGS.find_index(bearing)
    next_index = (current_index - 1) % BEARINGS.size
    next_bearing = BEARINGS[next_index]

    orient(next_bearing)
  end

  def orient(bearing)
    raise ArgumentError unless BEARINGS.include? bearing

    @bearing = bearing
  end

  def at(x, y)
    @coordinates = [x, y]
  end

  def advance
    x, y = coordinates

    case bearing
    when :east
      x += 1
    when :west
      x -= 1
    when :north
      y += 1
    when :south
      y -= 1
    end

    at(x, y)
  end

  private

  def initialize
    @coordinates = [0, 0]
  end
end

class Simulator
  def instructions(commands)
    commands.chars.map { command_to_instruction _1 }
  end

  def place(robot, x:, y:, direction:)
    robot.at(x, y)
    robot.orient(direction)
  end

  def evaluate(robot, commands)
    instructions(commands).each { robot.send _1 }
  end

  private

  def command_to_instruction(code)
    {
      'L' => :turn_left,
      'R' => :turn_right,
      'A' => :advance
    }[code]
  end
end
